[NNVM][TENSORFLOW] LSTM operator and PTB word prediction frontend

joyalbin opened the pull request:
This PR contain:
- [x] LSTM operator implementation and layer based operator parsing in Tensorflow
- [x] Single LSTMBlockCell layer model support
- [x] PTB LSTMBlockCell based model frontend script
- [x] Frontend testcases

joyalbin commented:
@Huyuwei @masahi @srkreddy1238 Please help to review this PR

srkreddy1238 commented on nnvm/python/nnvm/frontend/tensorflow.py:
```diff
@@ -469,6 +469,224 @@ def _impl(inputs, attr, params):
+    _, out_shapes = graph_util.infer_shape(g, **shape_dict)
+    return out_shapes
+
+def _stridedSlice():
```
Now we have strided_slice operator from nnvm. We should use it.

joyalbin replied:
@srkreddy1238 The frontend is using NNVM stridedslice only. But tensorflow have additional mask attributes for stridedslice. This is handled here.

srkreddy1238 commented on nnvm/python/nnvm/frontend/tensorflow.py:
```diff
@@ -469,6 +469,224 @@ def _impl(inputs, attr, params):
+                     'Taxis', '_class'])(new_input, attr)
+    return _impl
+
+def _infer_out_shapes(inputs, params):
```
We have already _input_shapes attribute. Any challenge using it?

joyalbin replied:
_infer_out_shapes is used for finding intermediate node shapes

srkreddy1238 commented on nnvm/tests/python/frontend/tensorflow/test_forward.py:
```diff
@@ -544,3 +711,6 @@ def test_forward_mobilenet():
     test_forward_inception_v1()
     test_forward_mobilenet()
     test_forward_variable()
+    test_forward_lstm()
```
Suggest to add a real time end to end test case for LSTM.

srkreddy1238 commented on tutorials/nnvm/from_tensorflow_rnn.py:
```diff
@@ -0,0 +1,242 @@
+"""
+Tutorial for Tensorflow RNN Models
```
underline up to the text length pls.

srkreddy1238 commented on tutorials/nnvm/from_tensorflow_rnn.py:
```diff
@@ -0,0 +1,242 @@
+sample_data_file = 'simple-examples.tgz'
+sample_url = sample_repo+sample_data_file
+
+ptb_repo = 'https://github.com/joyalbin/dmlc_store/raw/master/trained-models/tf/ptb/pb/'
```
use dmlc/web-data to store these model.

srkreddy1238 commented on tutorials/nnvm/from_tensorflow_rnn.py:
```diff
@@ -0,0 +1,242 @@
+
+###############################################################################
+# Input words
+# ---------------------------------------------
```
under line up to the text.

Huyuwei commented on nnvm/python/nnvm/frontend/tensorflow.py:
```diff
@@ -469,6 +469,224 @@ def _impl(inputs, attr, params):
+
+        o = sigmoid(cs * wco + o)
+        co = tanh(cs)
+        h = co .* o
```
can remove these math equations and give a link to https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/contrib/rnn/python/ops/lstm_ops.py#L41-L114

Huyuwei commented on nnvm/python/nnvm/frontend/tensorflow.py:
```diff
@@ -505,9 +723,203 @@ def _impl(inputs, attr, params):
+    """Recurrent network layer handlers.
+
+    Unlike normal operators, recurrent network operators have layer concept.
+    Same operators will be called multiple times (based on number of layers)
```
Does layer refer to steps? or multiple cells (layers) in stacked rnn? The doc here is not very clear.

joyalbin replied:
@Huyuwei here, layers refers to multiple cells in RNN stack. I have rephrased the comments

Huyuwei commented:
@joyalbin Seems current implementation only supports step=1 and users need to do unrolling manually in the case of multiple steps. Will it be good to add a wrapper to support multiple steps?

@merrymercy may have some suggestions.

joyalbin commented:
@Huyuwei , @srkreddy1238 I have reworked on all the review comments and updated the PR

@Huyuwei  Current implementation support step=1 at a time. This implementation is similar to tensorflow. Here the model input is one 'word', LSTM input is calculated from the 'word's token. so couldnt able to remove the unrolling.
Can we merge this PR with step=1 and multiple step handling in the next PR? Could you please put more light on avoiding this unrolling to handle multiple steps?

joyalbin commented:
@Huyuwei , @srkreddy1238 can you please help to review this PR further?

Huyuwei commented on nnvm/python/nnvm/frontend/tensorflow.py:
```diff
@@ -505,9 +701,198 @@ def _impl(inputs, attr, params):
+
+    Unlike normal operators, stacked rnn have cells and layer concepts.
+    Each Layer represent a cell in RNN stack. Cells in the same RNN stack
+    sequentialy process input data.
```
The description here is still a little confusing. Where is the recurrent part?
Can remove it or give a link to some RNN tutorial.

Huyuwei commented on nnvm/tests/python/frontend/tensorflow/test_forward.py:
```diff
@@ -528,6 +694,120 @@ def test_forward_mobilenet():
+    initializer = tf.random_uniform_initializer(config.init_scale,
+                                                config.init_scale)
+    with tf.variable_scope("Model", reuse=None, initializer=initializer):
+        mtest = nnvm.testing.tf.PTBModel(is_training=False, config=config)
```
mtest is not used? Then testing.tf.PTBModel can be removed

Huyuwei commented on nnvm/tests/python/frontend/tensorflow/test_forward.py:
```diff
@@ -528,6 +694,120 @@ def test_forward_mobilenet():
+    #TVM graph module creation
+    params, m = _get_tvm_graph_module(graph_def)
+
+
```
remove one blank line

Huyuwei commented:
@joyalbin Sorry for the late response. Have added some comments.

The lstm part looks good to me. fill, gather, stridedslice operators need review from @srkreddy1238

nishi-t commented on nnvm/python/nnvm/frontend/tensorflow.py:
```diff
@@ -505,9 +701,198 @@ def _impl(inputs, attr, params):
+            Dict of operator attributes
+
+        params : dict
+            List of  pretrained weights and bias
```
redundant whitespace after `of`

nishi-t commented on nnvm/python/nnvm/testing/tf.py:
```diff
@@ -257,3 +258,236 @@ def get_workload_mobilenet():
+    word_to_id : dict
+        English word to integer id mapping
+    id_to_word : dict
+
```
Could you remove this blank line?

nishi-t commented on nnvm/python/nnvm/frontend/tensorflow.py:
```diff
@@ -505,9 +701,198 @@ def _impl(inputs, attr, params):
+            Dict of operator attributes
+
+        params : dict
+            List of  pretrained weights and bias
```
whitespace after 'of'.

srkreddy1238 commented on nnvm/python/nnvm/frontend/tensorflow.py:
```diff
@@ -1,4 +1,4 @@
-# pylint: disable=import-self, invalid-name, unused-argument
+# pylint: disable=import-self, invalid-name, unused-argument, too-many-lines
```
too-many-lines ??
Suggest to have logical split instead. If any block can't be reduced use local enable/disable instead.

srkreddy1238 commented on nnvm/python/nnvm/frontend/tensorflow.py:
```diff
@@ -469,6 +469,211 @@ def _impl(inputs, attr, params):
+        new_axis_mask = int(attr.get('new_axis_mask', 0))
+        shrink_axis_mask = int(attr.get('shrink_axis_mask', 0))
+
+        #Constant values used forming output shape.
```
Thanks for cleaning the above logic.
Would suggest to clean the below part as well a bit.

My understanding of all these masks as below. Please correct me if I am wrong.

begin, end: Basically to ignore a value from the inputs and use 0 for begin, max for end.
elipsismask: Only one bit will be set which is used to expand begin, end and strides to size of input dimensions by filling max range in between.
newaxis mask: add new axis in the result based on the mask bits.
shringaxis_mask: shrink the axis from result based on the mask bits.

I see this logic can be split into below logically separable blocks.
1: Handle begin and end masks on begin, end
2: Expand begin and end based on elipsis_mask
3: Apply strided_slice operation
4: Apply reshape operation with newaxis and shrink axis 

Also suggest to use python API for converting bitmask into lists of integers.

srkreddy1238 commented on nnvm/tests/python/frontend/tensorflow/test_forward.py:
```diff
@@ -433,6 +441,155 @@ def test_forward_variable():
+    _test_stridedslice((3, 4, 3), [1, -1, 0], [4, -5, 3], [2, -1, 1], 'float32')
+    _test_stridedslice((3, 4, 3), [1, -3, 0], [2, -2, 3], [1, 1, 1], 'float32')
+    _test_stridedslice((3, 4, 3), [1, 1, 0], [4, 4, 3], [2, 1, 1], 'float32')
+    _test_stridedslice((3, 4, 3), [1, 0], [4, 3], [2, 1],
```
Please add another testcase with begin, end, ellipsis_mask and new/shrink axis mask with multiple axis.

joyalbin commented:
@srkreddy1238 yes, your understanding is correct. But handling all the 5 masks based on its priority make things bit lengthy.
I have modified the code based on your comments. 
Mask logic is changed as you suggested and testcases are added.

@Huyuwei I have handled all the review comments. Please help to approve the changes

Huyuwei commented on nnvm/python/nnvm/testing/tf.py:
```diff
@@ -134,3 +136,137 @@ def get_workload(model_path):
+    state = session.run(state_input_name)
+    fetches = [['Model/RNN/RNN/multi_rnn_cell/cell_0/lstm_cell/LSTMBlockCell:1',
+                'Model/RNN/RNN/multi_rnn_cell/cell_0/lstm_cell/LSTMBlockCell:6',
+                'Model/RNN/RNN/multi_rnn_cell/cell_0/lstm_cell/LSTMBlockCell_1:1',
```
Could you add a comment here? What is fetches? What is in the content of LSTMBlockCell:6?

Huyuwei commented on nnvm/python/nnvm/testing/tf.py:
```diff
@@ -134,3 +136,137 @@ def get_workload(model_path):
+    return int(np.searchsorted(t, 0.5 * s))
+
+def do_tf_sample(session, data, in_states, num_samples):
+    """Sampled from the model"""
```
``data`` is not used in the function.

joyalbin replied:
@Huyuwei `data `is used at line 192, could you please correct me if I didn't get the real issue here

Huyuwei replied:
@joyalbin sorry, my wrong.

tqchen commented:
Thanks @joyalbin @Huyuwei @nishi-t @srkreddy1238 , this is merged
